### PR TITLE
[SPARK-24050][SS] Calculate input / processing rates correctly for DataSourceV2 streaming sources

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

In some streaming queries, the input and processing rates are not calculated at all (shows up as zero) because MicroBatchExecution fails to associated metrics from the executed plan of a trigger with the sources in the logical plan of the trigger. The way this executed-plan-leaf-to-logical-source attribution works is as follows. With V1 sources, there was no way to identify which execution plan leaves were generated by a streaming source. So did a best-effort attempt to match logical and execution plan leaves when the number of leaves were same. In cases where the number of leaves is different, we just give up and report zero rates. An example where this may happen is as follows.

```
val cachedStaticDF = someStaticDF.union(anotherStaticDF).cache()
val streamingInputDF = ...

val query = streamingInputDF.join(cachedStaticDF).writeStream....
```
In this case, the `cachedStaticDF` has multiple logical leaves, but in the trigger's execution plan it only has leaf because a cached subplan is represented as a single InMemoryTableScanExec leaf. This leads to a mismatch in the number of leaves causing the input rates to be computed as zero. 

With DataSourceV2, all inputs are represented in the executed plan using `DataSourceV2ScanExec`, each of which has a reference to the associated logical `DataSource` and `DataSourceReader`. So its easy to associate the metrics to the original streaming sources.

In this PR, the solution is as follows. If all the streaming sources in a streaming query as v2 sources, then use a new code path where the execution-metrics-to-source mapping is done directly. Otherwise we fall back to existing mapping logic.

## How was this patch tested?
- New unit tests using V2 memory source
- Existing unit tests using V1 source

### CODE REVIEW

**Comment by tdas** [comment]
jenkins retest this please.

**Comment by tdas** [comment]
@brkyvz @jose-torres

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 1]
```diff
@@ -492,6 +492,77 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
...
+        assert(lastProgress.nonEmpty)
+        assert(lastProgress.get.numInputRows == 3)
+        assert(lastProgress.get.sources.length == 1)
+        assert(lastProgress.get.sources(0).numInputRows == 3)
+        true
+      }
+    )
+
+    val streamInput2 = MemoryStream[Int]
+    val staticInputDF2 = staticInputDF.union(staticInputDF).cache()
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: nit: unpersist later?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 2]
```diff
@@ -733,6 +804,11 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
     }
   }
 
+  def getLastProgressWithData(q: StreamingQuery): Option[StreamingQueryProgress] = {
+    q.recentProgress.filter(_.numInputRows > 0).lastOption
+  }
+
+
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: nit: extra line

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 3]
```diff
@@ -492,6 +492,77 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
...
+        assert(lastProgress.get.sources.length == 1)
+        assert(lastProgress.get.sources(0).numInputRows == 3)
+        true
+      }
+    )
+
+    val streamInput2 = MemoryStream[Int]
+    val staticInputDF2 = staticInputDF.union(staticInputDF).cache()
+
+    testStream(streamInput2.toDF().join(staticInputDF2, "value"), useV2Sink = true)(
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: what if you do a stream-stream join?

Comment (brkyvz) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: e.g. self-join?

**Review by brkyvz — COMMENTED** [review]
What is the behavior if there's a self join? Will we double count the numInputRows?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala (hunk 1 above)
Comment (tdas) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: really doesnt matter as the testsuite will shutdown the sparkcontext anyways.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala (hunk 3 above)
Comment (tdas) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: then there will be two DataSourceScanV2Execs reading from the same location. So we will be reading data twice, and the counts will reflect that. But yes, I should add a test for that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala [hunk 4]
```diff
@@ -207,62 +209,92 @@ trait ProgressReporter extends Logging {
...
     }
-    val allLogicalPlanLeaves = lastExecution.logical.collectLeaves() // includes non-streaming
-    val allExecPlanLeaves = lastExecution.executedPlan.collectLeaves()
-    val numInputRows: Map[BaseStreamingSource, Long] =
+
+    val onlyDataSourceV2Sources = {
+      // Check whether the streaming query's logical plan has only V2 data sources
+      val allStreamingLeaves =
+        logicalPlan.collect { case s: StreamingExecutionRelation => s }
+      allStreamingLeaves.forall { _.source.isInstanceOf[MicroBatchReader] }
```
[the remark this answers is not in the document]
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala]: Maybe i can make it work for continuous as well with a small tweak

[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala]: A point fix here won't be sufficient - right now the metrics don't make it to the driver at all in continuous processing.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 5]
```diff
@@ -492,6 +492,77 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
...
+      CheckAnswer(1, 2, 3),
+      AssertOnQuery { q =>
+        val lastProgress = getLastProgressWithData(q)
+        assert(lastProgress.nonEmpty)
+        assert(lastProgress.get.numInputRows == 3)
+        assert(lastProgress.get.sources.length == 2)
+        assert(lastProgress.get.sources(0).numInputRows == 3)
+        assert(lastProgress.get.sources(1).numInputRows == 0)
+        true
+      }
```
Comment (jose-torres) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: nit: i'd suggest doing an AddData() for the other stream after, to make sure there's not some weird order dependence

**Comment by jose-torres** [comment]
lgtm

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala [hunk 6]
```diff
@@ -207,62 +209,126 @@ trait ProgressReporter extends Logging {
...
       } else {
         if (!metricWarningLogged) {
           def toString[T](seq: Seq[T]): String = s"(size = ${seq.size}), ${seq.mkString(", ")}"
+
           logWarning(
             "Could not report metrics as number leaves in trigger logical plan did not match that" +
-                s" of the execution plan:\n" +
-                s"logical plan leaves: ${toString(allLogicalPlanLeaves)}\n" +
-                s"execution plan leaves: ${toString(allExecPlanLeaves)}\n")
+              s" of the execution plan:\n" +
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala]: existing nit: maybe we should've just used 
```
"""
   |
   |
"""
```

**Review by brkyvz — APPROVED** [review]
LGTM. Thanks for all the tests. I'm trusting the unit tests. I don't see a better way of figuring out unique sources.

**Comment by tdas** [comment]
jenkins retest this please

**Comment by tdas** [comment]
jenkins retest this please

**Comment by tdas** [comment]
jenkins retest this please

**Pull request closed without merging**